### PR TITLE
handle 401 error during plugin setup

### PR DESCRIPTION
**Pull request by tianleh**
*Issue #, if available:*

*Description of changes:*

Fix the issue that 401 error is not handled.

*Test*

Before the change,
Screen Shot 2020-08-06 at 1 18 51 PM

After the change,
Screen Shot 2020-08-06 at 1 24 32 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/plugin.ts [hunk 1]
```diff
@@ -35,7 +35,17 @@ export class OpendistroSecurityPlugin
...
   public async setup(core: CoreSetup): Promise<OpendistroSecurityPluginSetup> {
-    const permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    let permissions;
+
+    try {
+      permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    } catch (e) {
+      if (e.body.statusCode === 401) {
+        // ignore and continue to login page.
+      } else {
```
Comment (yuxitang-amzn) [inline public/plugin.ts:45]: Can be simplified to statusCode !== 401

File: public/plugin.ts [hunk 2]
```diff
@@ -35,7 +35,17 @@ export class OpendistroSecurityPlugin
...
+        throw e;
+      }
+    }
 
     if (permissions?.data?.has_api_access) {
```
Comment (yuxitang-amzn) [inline public/plugin.ts:50]: Suggest extract to separate function so easier to add unit test later.

File: public/plugin.ts [hunk 3]
```diff
@@ -35,7 +35,17 @@ export class OpendistroSecurityPlugin
   constructor(private readonly initializerContext: PluginInitializerContext) {}
 
   public async setup(core: CoreSetup): Promise<OpendistroSecurityPluginSetup> {
-    const permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    let permissions;
+
+    try {
+      permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    } catch (e) {
+      if (e.body.statusCode === 401) {
```
Comment (zengyan-amazon) [inline public/plugin.ts:43]: maybe `e.body?.statusCode` to avoid referring field of undefined?

Comment (tianleh) [reply public/plugin.ts]: Sure

File: public/plugin.ts (hunk 1 above)
Comment (tianleh) [reply public/plugin.ts]: Sure

File: public/plugin.ts (hunk 2 above)
Comment (tianleh) [reply public/plugin.ts]: ok

File: public/plugin.ts [hunk 4]
```diff
@@ -29,15 +29,27 @@ import {
...
+}
+
 export class OpendistroSecurityPlugin
   implements Plugin<OpendistroSecurityPluginSetup, OpendistroSecurityPluginStart> {
   // @ts-ignore : initializerContext not used
   constructor(private readonly initializerContext: PluginInitializerContext) {}
 
   public async setup(core: CoreSetup): Promise<OpendistroSecurityPluginSetup> {
-    const permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    const apiPermission = await getApiPermission(core);
```
Comment (yuxitang-amzn) [inline public/plugin.ts:50]: Minor: suggest we have a more boolean-like name like `hasApiPermission`

File: public/plugin.ts [hunk 5]
```diff
@@ -29,15 +29,27 @@ import {
 import { LOGIN_PAGE_URI, PLUGIN_NAME } from '../common';
 import { API_ENDPOINT_PERMISSIONS_INFO } from './apps/configuration/constants';
 
+async function getApiPermission(core: CoreSetup): Promise<boolean | undefined> {
+  try {
+    const permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
+    return permissions?.data?.has_api_access;
```
Comment (yuxitang-amzn) [inline public/plugin.ts:35]: Minor: we can do `|| false` so the return type could be just `Promise<boolean>`

File: public/plugin.ts (hunk 4 above)
Comment (tianleh) [reply public/plugin.ts]: Sure

File: public/plugin.ts (hunk 5 above)
Comment (tianleh) [reply public/plugin.ts]: Added || false. But tslint check shows that undefined is still needed because of the exception throwing.

**Pull request merged**